File: src/main/java/com/codingforcookies/betterrecords/block/BlockSpeaker.java
package com.codingforcookies.betterrecords.block;

import com.codingforcookies.betterrecords.api.wire.IRecordWire;
import com.codingforcookies.betterrecords.client.core.handler.BetterEventHandler;
import com.codingforcookies.betterrecords.block.tile.TileSpeaker;
import com.codingforcookies.betterrecords.common.core.handler.ConfigHandler;
import com.codingforcookies.betterrecords.common.core.helper.ConnectionHelper;
import com.codingforcookies.betterrecords.common.util.BetterUtils;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;

public class BlockSpeaker extends BetterBlock {

    public static String[] speakers = new String[]{"sm", "md", "lg"};
    public int meta = 0;

    public BlockSpeaker(String name, int meta){
        super(Material.WOOD, name);
        this.meta = meta;

        switch (meta) {
            case 0: setHardness(2F); setResistance(7.5F); break;
            case 1: setHardness(3F); setResistance(8F);   break;
            case 2: setHardness(4F); setResistance(9.5F); break;
        }
    }

    @Override
    public void onBlockAdded(World world, BlockPos pos, IBlockState state){
        super.onBlockAdded(world, pos, state);
        world.notifyBlockUpdate(pos, state, state, 3);
    }

    @Override
    public AxisAlignedBB getBoundingBox(IBlockState state, IBlockAccess block, BlockPos pos) {
        switch (meta){
            case 0:
                return new AxisAlignedBB(0.26F, 0.05F, 0.25F, 0.75F, 0.65F, 0.74F);
            case 1:
                return new AxisAlignedBB(0.2F, 0.0F, 0.2F, 0.8F, 0.88F, 0.8F);
            case 2:
                return new AxisAlignedBB(0.12F, 0.0F, 0.12F, 0.88F, 1.51F, 0.88F);
            default:
                return FULL_BLOCK_AABB;
        }
    }

    @Override
    public void onBlockPlacedBy(World world, BlockPos pos, IBlockState state, EntityLivingBase placer, ItemStack stack) {
        TileEntity tileEntity = world.getTileEntity(pos);
        if(tileEntity == null || !(tileEntity instanceof TileSpeaker)) return;
        ((TileSpeaker) tileEntity).rotation = placer.rotationYaw;
        ((TileSpeaker) tileEntity).type = meta;
        if(world.isRemote && !ConfigHandler.tutorials.get("speaker")) {
            BetterEventHandler.tutorialText = BetterUtils.getTranslatedString("tutorial.speaker");
            BetterEventHandler.tutorialTime = System.currentTimeMillis() + 10000;
            ConfigHandler.tutorials.put("speaker", true);
        }
    }

    @Override
    public boolean removedByPlayer(IBlockState state, World world, BlockPos pos, EntityPlayer player, boolean willHarvest){
        if(world.isRemote) return super.removedByPlayer(state, world, pos, player, willHarvest);
        TileEntity te = world.getTileEntity(pos);
        if(te != null && te instanceof IRecordWire) ConnectionHelper.clearConnections(world, (IRecordWire) te);
        return super.removedByPlayer(state, world, pos, player, willHarvest);
    }

    @Override
    public TileEntity createNewTileEntity(World var1, int var2){
        return new TileSpeaker();
    }
}

File: src/main/kotlin/com/codingforcookies/betterrecords/block/BlockSpeaker.kt
package com.codingforcookies.betterrecords.block

import com.codingforcookies.betterrecords.ID
import com.codingforcookies.betterrecords.api.wire.IRecordWire
import com.codingforcookies.betterrecords.block.tile.TileSpeaker
import com.codingforcookies.betterrecords.client.render.RenderSpeaker
import com.codingforcookies.betterrecords.helper.ConnectionHelper
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.block.properties.PropertyEnum
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.IStringSerializable
import net.minecraft.util.NonNullList
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import net.minecraftforge.client.ForgeHooksClient
import net.minecraftforge.client.model.ModelLoader

class BlockSpeaker(name: String) : ModBlock(Material.WOOD, name), TESRProvider<TileSpeaker>, ItemModelProvider {

    companion object {
        val PROPERTYSIZE = PropertyEnum.create("size", SpeakerSize::class.java)
    }

    init {
        setHardness(when (0) {
            0 -> 2F
            1 -> 3F
            2 -> 4F
            else -> 2F // Uh oh
        })

        setResistance(when (0) {
            0 -> 7.5F
            1 -> 8F
            2 -> 9.5F
            else -> 7.5F // Uh oh
        })
    }

    override fun getTileEntityClass() = TileSpeaker::class
    override fun getRenderClass() = RenderSpeaker::class

    override fun getBoundingBox(state: IBlockState, block: IBlockAccess?, pos: BlockPos?) = getBouningBoxFromState(state)
    override fun getCollisionBoundingBox(state: IBlockState, worldIn: IBlockAccess?, pos: BlockPos?) = getBouningBoxFromState(state)

    private fun getBouningBoxFromState(state: IBlockState) = when (state.getValue(PROPERTYSIZE)) {
        SpeakerSize.SMALL -> AxisAlignedBB(0.26, 0.05, 0.25, 0.75, 0.65, 0.74)
        SpeakerSize.MEDIUM -> AxisAlignedBB(0.2, 0.0, 0.2, 0.8, 0.88, 0.8)
        SpeakerSize.LARGE -> AxisAlignedBB(0.12, 0.0, 0.12, 0.88, 1.51, 0.88)
        else -> Block.FULL_BLOCK_AABB
    }

    override fun onBlockAdded(world: World, pos: BlockPos, state: IBlockState) =
            world.notifyBlockUpdate(pos, state, state, 3)

    override fun onBlockPlacedBy(world: World, pos: BlockPos, state: IBlockState, placer: EntityLivingBase, stack: ItemStack) {
        (world.getTileEntity(pos) as? TileSpeaker)?.let { te ->
            te.rotation = placer.rotationYaw
            te.size = state.getValue(PROPERTYSIZE)
        }
    }

    override fun removedByPlayer(state: IBlockState, world: World, pos: BlockPos, player: EntityPlayer, willHarvest: Boolean): Boolean {
        if (!world.isRemote) {
            (world.getTileEntity(pos) as? IRecordWire)?.let { te ->
                ConnectionHelper.clearConnections(world, te)
            }
        }
        return super.removedByPlayer(state, world, pos, player, willHarvest)
    }

    override fun damageDropped(state: IBlockState) = state.getValue(PROPERTYSIZE).meta

    override fun getSubBlocks(itemIn: CreativeTabs, items: NonNullList<ItemStack>) {
        SpeakerSize.values().mapTo(items) {
            ItemStack(this, 1, it.meta)
        }
    }

    override fun getStateFromMeta(meta: Int) = this.defaultState.withProperty(PROPERTYSIZE, SpeakerSize.fromMeta(meta))
    override fun getMetaFromState(state: IBlockState) = state.getValue(PROPERTYSIZE).meta

    override fun createBlockState() = BlockStateContainer(this, PROPERTYSIZE)

    override fun registerItemModel(block: ModBlock) {
        val item = Item.getItemFromBlock(block)
        ModelLoader.setCustomModelResourceLocation(item, 0, ModelResourceLocation("$ID:itemblock/$name", "inventory"))
        ModelLoader.setCustomModelResourceLocation(item, 1, ModelResourceLocation("$ID:itemblock/$name", "inventory"))
        ModelLoader.setCustomModelResourceLocation(item, 2, ModelResourceLocation("$ID:itemblock/$name", "inventory"))
    }

    override fun registerTESRItemStacks(block: ModBlock) {
        val item = Item.getItemFromBlock(block)
        ForgeHooksClient.registerTESRItemStack(item, 0, getTileEntityClass().java)
        ForgeHooksClient.registerTESRItemStack(item, 1, getTileEntityClass().java)
        ForgeHooksClient.registerTESRItemStack(item, 2, getTileEntityClass().java)
    }

    enum class SpeakerSize(val meta: Int, val typeName: String) : IStringSerializable {
        SMALL(0, "small"),
        MEDIUM(1, "medium"),
        LARGE(2, "large");

        override fun getName() = typeName

        companion object {
            private val map = SpeakerSize.values().associateBy(SpeakerSize::meta)
            fun fromMeta(meta: Int) = map[meta] ?: SMALL
        }
    }
}
